File: packages/python/semantic_layer/parent_context.py
"""Small W3C/active-OpenTelemetry parent-context resolver with no global takeover."""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Literal


@dataclass(frozen=True)
class ResolvedParentContext:
    traceparent: str | None = None
    gap: Literal[
        "required_parent_context_missing",
        "invalid_traceparent",
        "parent_context_unreadable",
        "parent_context_conflict",
    ] | None = None
    error: BaseException | None = None


def resolve_parent_context(
    explicit: object | None, inherited: str | None = None
) -> ResolvedParentContext:
    try:
        if isinstance(explicit, Mapping):
            value = explicit.get("traceparent")
            required = explicit.get("required") is True
        elif explicit is None:
            value = None
            required = False
        else:
            value = explicit
            required = False
    except BaseException as error:
        return ResolvedParentContext(
            traceparent=inherited, gap="parent_context_unreadable", error=error
        )
    if value is not None:
        if not is_valid_traceparent(value):
            return ResolvedParentContext(traceparent=inherited, gap="invalid_traceparent")
        if inherited is not None and value != inherited:
            return ResolvedParentContext(
                traceparent=inherited, gap="parent_context_conflict"
            )
        return ResolvedParentContext(traceparent=inherited or value)
    if inherited is not None:
        return ResolvedParentContext(traceparent=inherited)
    active = _active_otel_traceparent()
    if active is not None:
        return ResolvedParentContext(traceparent=active)
    return (
        ResolvedParentContext(gap="required_parent_context_missing")
        if required
        else ResolvedParentContext()
    )


def is_valid_traceparent(value: object) -> bool:
    if not isinstance(value, str) or len(value) > 512:
        return False
    match = re.fullmatch(
        r"([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})(.*)",
        value,
    )
    if match is None:
        return False
    version, trace_id, parent_id, _flags, future = match.groups()
    if version == "ff":
        return False
    if set(trace_id) == {"0"}:
        return False
    if set(parent_id) == {"0"}:
        return False
    if version == "00":
        return not future
    return not future or future.startswith("-")


def _active_otel_traceparent() -> str | None:
    try:
        from opentelemetry.trace import get_current_span

        context: Any = get_current_span().get_span_context()
        if not bool(getattr(context, "is_valid", False)):
            return None
        trace_id = int(context.trace_id)
        span_id = int(context.span_id)
        flags = int(context.trace_flags) & 0xFF
        candidate = f"00-{trace_id:032x}-{span_id:016x}-{flags:02x}"
        return candidate if is_valid_traceparent(candidate) else None
    except Exception:
        return None

File: packages/sdk/src/v1/parent-context.ts
import { createRequire } from 'node:module';
import type { ParentContext } from './types.js';

export type ResolvedParentContext = Readonly<{
  traceparent?: string;
  gap?: 'required_parent_context_missing' | 'invalid_traceparent' | 'parent_context_unreadable' | 'parent_context_conflict';
  error?: unknown;
}>;

type OTelApi = {
  context?: { active(): unknown };
  trace?: {
    getSpanContext(value: unknown): {
      traceId?: unknown;
      spanId?: unknown;
      traceFlags?: unknown;
      isRemote?: unknown;
    } | undefined;
  };
};

const require = createRequire(import.meta.url);
let otelApi: OTelApi | null | undefined;

/** Resolve explicit, inherited, or active-OTel context without installing global instrumentation. */
export function resolveParentContext(
  explicit: ParentContext | undefined,
  inherited?: string,
): ResolvedParentContext {
  const value = readParentContext(explicit);
  if (value.unreadable) return { ...withGap(inherited, 'parent_context_unreadable'), error: value.error };
  if (value.traceparent !== undefined) {
    if (!isValidTraceparent(value.traceparent)) return withGap(inherited, 'invalid_traceparent');
    if (inherited !== undefined && value.traceparent !== inherited) {
      return { traceparent: inherited, gap: 'parent_context_conflict' };
    }
    return { traceparent: inherited ?? value.traceparent };
  }
  if (inherited !== undefined) return { traceparent: inherited };
  const active = activeOTelTraceparent();
  if (active !== undefined) return { traceparent: active };
  return value.required ? { gap: 'required_parent_context_missing' } : {};
}

/** Read the public option without letting a hostile getter interrupt customer code. */
export function resolveParentContextOption(
  options: unknown,
  inherited?: string,
): ResolvedParentContext {
  try {
    if (!options || typeof options !== 'object') return resolveParentContext(undefined, inherited);
    return resolveParentContext(Reflect.get(options, 'parentContext') as ParentContext | undefined, inherited);
  } catch (error) {
    return { ...withGap(inherited, 'parent_context_unreadable'), error };
  }
}

export function isValidTraceparent(value: unknown): value is string {
  if (typeof value !== 'string' || value.length > 512) return false;
  const match = /^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})(.*)$/.exec(value);
  if (!match) return false;
  const [, version, traceId, parentId, , future] = match;
  if (version === 'ff' || /^0+$/.test(traceId) || /^0+$/.test(parentId)) return false;
  if (version === '00') return future.length === 0;
  return future.length === 0 || future.startsWith('-');
}

function readParentContext(value: ParentContext | undefined): {
  traceparent?: unknown;
  required: boolean;
  unreadable: boolean;
  error?: unknown;
} {
  if (value === undefined) return { required: false, unreadable: false };
  if (!value || typeof value !== 'object') return { traceparent: value, required: false, unreadable: false };
  try {
    return {
      traceparent: Reflect.get(value, 'traceparent'),
      required: Reflect.get(value, 'required') === true,
      unreadable: false,
    };
  } catch (error) {
    return { required: false, unreadable: true, error };
  }
}

function withGap(inherited: string | undefined, gap: NonNullable<ResolvedParentContext['gap']>): ResolvedParentContext {
  return inherited === undefined ? { gap } : { traceparent: inherited, gap };
}

function activeOTelTraceparent(): string | undefined {
  try {
    if (otelApi === undefined) otelApi = require('@opentelemetry/api') as OTelApi;
    if (otelApi === null) return undefined;
    const active = otelApi.context?.active();
    const span = active === undefined ? undefined : otelApi.trace?.getSpanContext(active);
    const traceId = span?.traceId;
    const spanId = span?.spanId;
    const traceFlags = span?.traceFlags;
    if (typeof traceId !== 'string' || typeof spanId !== 'string') return undefined;
    const flags = typeof traceFlags === 'number' ? (traceFlags & 0xff).toString(16).padStart(2, '0') : '00';
    const candidate = `00-${traceId}-${spanId}-${flags}`;
    return isValidTraceparent(candidate) ? candidate : undefined;
  } catch {
    otelApi = null;
    return undefined;
  }
}
